Python/Product/Analysis/Infrastructure/StringExtensions.cs

```diff
@@ -16,15 +16,68 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
 using System.Linq;
+using System.Text.RegularExpressions;
 
 namespace Microsoft.PythonTools.Analysis.Infrastructure {
     static class StringExtensions {
+#if DEBUG
+        private static readonly Regex SubstitutionRegex = new Regex(
+            @"\{(\d+)",
+            RegexOptions.IgnorePatternWhitespace,
+            TimeSpan.FromSeconds(1)
+        );
+
+        private static void ValidateFormatString(string str, int argCount) {
+            foreach (Match m in SubstitutionRegex.Matches(str)) {
+                int index = int.Parse(m.Groups[1].Value, CultureInfo.InvariantCulture);
+                if (index >= argCount) {
+                    Debug.Fail(string.Format(CultureInfo.InvariantCulture, "Format string expects more than {0} args.\n\n{1}", argCount, str));
```

Comment:
`string.Format` already throws `FormatException` when string expects more arguments than provided. Why do you need to check manually?

Reply:
The Debug.Fail will show up more prominently while we're working (FormatException will likely be swallowed and logged somewhere, which adds a few more steps to find it, and this also only reports mismatched argument count rather than other formatting errors).

It's also leftover from when we used to check for unused arguments as well. I don't remember exactly why we took that out, but I do remember it caught a lot of bugs. @huguesv - any memory of this?

Reply:
sorry, no memory